internal/pkg/v2/infrastructure/redis/provisionwatcher.go

```diff
@@ -0,0 +1,81 @@
+//
+// Copyright (C) 2021 IOTech Ltd
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package redis
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/edgexfoundry/edgex-go/internal/pkg/common"
+	"github.com/edgexfoundry/go-mod-core-contracts/errors"
+	v2 "github.com/edgexfoundry/go-mod-core-contracts/v2"
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/models"
+
+	"github.com/gomodule/redigo/redis"
+)
+
+const (
+	ProvisionWatcherCollection            = "md|pw"
+	ProvisionWatcherCollectionName        = ProvisionWatcherCollection + DBKeySeparator + v2.Name
+	ProvisionWatcherCollectionLabel       = ProvisionWatcherCollection + DBKeySeparator + v2.Label
+	ProvisionWatcherCollectionServiceName = ProvisionWatcherCollectionName + DBKeySeparator + v2.Service + DBKeySeparator + v2.Name
+	ProvisionWatcherCollectionProfileName = ProvisionWatcherCollectionName + DBKeySeparator + v2.Profile + DBKeySeparator + v2.Name
+)
+
+// provisionWatcherStoredKey return the provision watcher's stored key which combines the collection name and object id
+func provisionWatcherStoredKey(id string) string {
+	return CreateKey(ProvisionWatcherCollection, id)
+}
+
+// addProvisionWatcher adds a new provision watcher into DB
+func addProvisionWatcher(conn redis.Conn, pw models.ProvisionWatcher) (addedProvisionWatcher models.ProvisionWatcher, edgexErr errors.EdgeX) {
+	// retrieve provision watcher by Id first to ensure there is no Id conflict; when Id exists, return duplicate error
+	exists, edgexErr := objectIdExists(conn, provisionWatcherStoredKey(pw.Id))
+	if edgexErr != nil {
+		return addedProvisionWatcher, errors.NewCommonEdgeXWrapper(edgexErr)
+	} else if exists {
+		return addedProvisionWatcher, errors.NewCommonEdgeX(errors.KindDuplicateName, fmt.Sprintf("provision watcher id %s already exists", pw.Id), edgexErr)
+	}
+
+	// verify if provision watcher name is unique or not
+	exists, edgexErr = objectNameExists(conn, ProvisionWatcherCollectionName, pw.Name)
+	if edgexErr != nil {
+		return addedProvisionWatcher, errors.NewCommonEdgeXWrapper(edgexErr)
+	} else if exists {
+		return addedProvisionWatcher, errors.NewCommonEdgeX(errors.KindDuplicateName, fmt.Sprintf("provision watcher name %s already exists", pw.Name), edgexErr)
+	}
+
+	ts := common.MakeTimestamp()
+	// For Redis DB, the PUT or PATCH operation will removes the old object and add the modified one,
```

Comment:
This comment indicates that this func will also be used for PUT or PATCh, but can't do to above checks. Please rework the comment.

Reply:
gonna remove the comment here.